keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,141 @@
+
+  Currently `kubectl apply --prune` is still in Alpha. It is based on label matching. Users need to be really careful to use this command, so that they don't delete objects unintentionally. There is a list of [issues]([kubernetes/kubectl#572]) observed by users. The most noticeable issue happens when changing directories or working with directories with the same labels. When two directories share the same label. `kubectl apply --prune` from one directory will delete objects from the other directory.
+
+ In this design, we take an overview of the pruning challenges and propose a better pruning solution. This new approach is able to work with or to be utilized by generic configuration management tools.
+
+### Goals
+
+Propose a solution for better `kubectl apply --prune` so that
+- it can find the correct set of objects to delete
+- it can avoid mess up when changing directories
+
+### Non-Goals
+
+- Make `--prune` default on
+- Improvement on Garbage Collections
+
+## Proposal
+
+### Challenges
+The challenges of pruning are mainly from two aspects.
+- Find out correct objects to delete. When a directory is applied after some updates, it loses the track of objects that are previously applied from the same directory.
+- Delete objects at proper time. Assume the set of objects to be deleted is available, when to delete them is another challenge. Those objects could be still needed during rolling out other objects.
+
+In this design, we focus on a solution for the first challenge. For the second challenge, we can blacklist certain types such as ConfigMaps and Secrets for deletion. Generally deleting objects at proper time can be better addressed by Garbage Collection.
+
+
+### Root Object
+To solve the first challenge, a concept of root object is proposed. A root object is a ConfigMap object that contains the GVKN information for all the objects that are to be applied. The root ConfigMap objects should be created by other tools such as kustomize, Helm.  When generating resources to be consumed by kubectl apply, those tools are responsible to generate the associated root ConfigMap object to utilize better pruning.
+
+Here is an example of root ConfigMap object. It has a predictable name or a label `kubectl.kubernetes.io/applyRoot: true` indicating that it is a root object.
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: predictable-name-foo
+ labels:
+   kubectl.kubernetes.io/applyRoot: true
+data:
+ object1:  namespace1:group1:version1:kind1:somename
+ object2:  namespace1:group2:version2:kind2:secondname
+ object3:  namespace3:group3:version3:kind3:thirdname
+```
+
+The presence of a root object indicates apply need to handle prune. When kubectl apply is given such a root object, it finds the live root object with the same name. 
```

Comment:
Do you mean: When kubectl apply is given _name of_ a root object? For example: `kubectl apply -f --prune -r <some-root-object-name>` (adding `-r` flag for inputting the root object name).

Reply:
We don't introduce new flags for this. It will be `kubectl apply --prune -f`. As long as a ConfigMap object is present in the resources passed by `-f` with a proper lable, Kubectl will treat it as a root object.

Reply:
Got it, thanks!